File: C++/题目1414：旅游啦.cpp

#include <stdio.h>
#include <string.h>
#define INF 10000000;
const int maxn = 50;

int path[maxn][maxn];
int minLen[maxn];
int visited[maxn];
int m,i,j;
int dijkstra(){
	minLen[0] = 0;
	visited[0] = 1;
	int minj = 0;
	for(i = 0; i < maxn; i++){
		int min = INF;
		for(j = 0; j < maxn; j++){
			if(visited[j] == 0 && minLen[j] < min){
				min = minLen[j];
				minj = j;
			}
		}
		visited[minj] = 1;
		for(j = 0; j < maxn ; j++){
			if (visited[j] == 0 && minLen[j] > (minLen[minj] + path[minj][j])) {
				minLen[j] = minLen[minj] + path[minj][j];
			}
		}
	}
	int maxLen = 0;
	for (i = 1; i < maxn; i++) {
	    if(maxLen < minLen[i] && minLen[i] < 10000000){
			maxLen = minLen[i];
	    }
	}
	return maxLen;
}
int main(){
	while(scanf("%d",&m) != EOF){
		for(i = 0; i < maxn; i++){
			for(j = 0; j < maxn; j++){
				path[i][j] = INF;
			}
		}
		int sum = 0;
		for(i = 0; i < m; i++){
			int a,b,d;
			scanf("%d%d%d",&a,&b,&d);
			if(path[a][b] > d){
				path[a][b] =d;
				path[b][a] = d;
				sum += d;
			}
		}
		memset(visited,0,sizeof(visited));
		for(i = 1; i < maxn; i++){
			minLen[i] = path[0][i];
		}
		int maxLen = dijkstra();
		printf("%d\n",2*sum - maxLen);
	}
	return 0;
}

/**************************************************************
	Problem: 1414
	User: wangzhenqing
	Language: C++
	Result: Accepted
	Time:30 ms
	Memory:1032 kb
****************************************************************/


                        

File: Java/题目1414：旅游啦.java

// 题目1414：旅游啦

/**
 * @author:wangzq
 * @email:wangzhenqing1008@163.com
 * @date:2015-06-30 11:01:55
 * @url:http://ac.jobdu.com/problem.php?pid=1414
 * 解题思路参考csdn:http://blog.csdn.net/u013027996/article/details/17576961
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.StreamTokenizer;

public class Main {
    /*
     * 1414
     */
    private static int n = 50;

    public static void main(String[] args) throws Exception {
        StreamTokenizer st = new StreamTokenizer(new BufferedReader(
                new InputStreamReader(System.in)));
        while (st.nextToken() != StreamTokenizer.TT_EOF) {
            int m = (int) st.nval;
            int cost[][] = new int[n][n];
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    cost[i][j] = Integer.MAX_VALUE;
                }
            }
            int sum = 0;
            for (int i = 0; i < m; i++) {
                st.nextToken();
                int a = (int) st.nval;
                st.nextToken();
                int b = (int) st.nval;
                st.nextToken();
                int d = (int) st.nval;
                if (cost[a][b] > d) {
                    cost[a][b] = d;
                    cost[b][a] = d;
                    sum += d;
                }
            }
            int minCost[] = new int[n];
            int visit[] = new int[n];

            for (int i = 0; i < n; i++) {
                minCost[i] = cost[0][i];
            }
            int maxLen = dijkstra(cost, minCost, visit);
            System.out.printf("%d\n", 2 * sum - maxLen);
        }
    }

    private static int dijkstra(int[][] cost, int[] minCost, int[] visit) {
        minCost[0] = 0;
        visit[0] = 1;
        int minj = 0;
        for (int i = 0; i < n; i++) {
            int min = Integer.MAX_VALUE;
            for (int j = 0; j < n; j++) {
                if (visit[j] == 0 && minCost[j] < min) {
                    min = minCost[j];
                    minj = j;
                }
            }
            visit[minj] = 1;
            for (int j = 0; j < n; j++) {
                if (visit[j] == 0
                        && minCost[j] > (minCost[minj] + cost[minj][j])
                        && minCost[minj] != Integer.MAX_VALUE
                        && cost[minj][j] != Integer.MAX_VALUE) {
                    minCost[j] = minCost[minj] + cost[minj][j];
                }
            }
        }
        int maxLen = 0;
        for (int i = 0; i < n; i++) {
            if (maxLen < minCost[i] && minCost[i] != Integer.MAX_VALUE) {
                maxLen = minCost[i];
            }
        }
        return maxLen;
    }
}
/**************************************************************
 Problem: 1414
 User: wzqwsrf
 Language: Java
 Result: Accepted
 Time:400 ms
 Memory:24856 kb
 ****************************************************************/




